v2/src/logger.c

```diff
@@ -0,0 +1,49 @@
+// Copyright (c) Microsoft. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+#include <stdint.h>
+#include <stdarg.h>
+#include <stdio.h>
+
+#include "c_logging/logger.h"
+#include "c_logging/log_sink.h"
+#include "c_logging/log_sink_console.h"
+#include "c_logging/log_sink_etw.h"
+#include "c_logging/log_context.h"
+
+MU_DEFINE_ENUM_STRINGS_WITHOUT_INVALID(LOG_LEVEL, LOG_LEVEL_VALUES);
+
+const LOG_SINK* log_sinks[] = { &console_log_sink, &etw_log_sink, NULL };
+const char logging_error_message[] = "Error generating log message";
+
+void logger_log(LOG_LEVEL log_level, LOG_CONTEXT_HANDLE log_context, const char* file, const char* func, int line_no, const char* format, ...)
+{
+    uint32_t i;
+    va_list args;
+    va_start(args, format);
+
+    char message[LOG_MAX_MESSAGE_LENGTH];
+
+    int result = vsnprintf(message, LOG_MAX_MESSAGE_LENGTH, format, args);
+    if (result < 0)
+    {
+        (void)memcpy(message, logging_error_message, sizeof(logging_error_message));
+    }
+    else if (result == LOG_MAX_MESSAGE_LENGTH)
```

Comment:
>LOG_MAX_MESSAGE_LENGTH [](http://example.com/codeflow?start=23&length=22)

I am not sure that C standard says that vsnprintf will deploy any characters in the destitnation if it returns LOG_MAX_MESSAGE_LENGT.


Or I don't remember,...

Reply:
The vsnprintf function returns the number of characters that would have been written
had n been sufficiently large, not counting the terminating null character, oraneg ative
value if an encoding error occurred. Thus, the null-terminated output has been
completely written if and only if the returned value is nonnegative and less than n.

The way I read this if LOG_MAX_MESSAGE_LENGTH the null terminator is not written.